pkg/local-pv-monitor/monitor.go

```diff
@@ -0,0 +1,508 @@
+	if !dir && !bl && (dirErr != nil || blErr != nil) {
+		// mountPath does not exist or is not a directory
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, HostPathNotExist, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	continueThisCheck = true
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, HostPathNotExist, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+
+}
+
+func (monitor *LocalPVMonitor) checkPVAndFSSize(mountPath string, pv *v1.PersistentVolume) {
+	capacityByte, err := monitor.VolUtil.GetFsCapacityByte(mountPath)
+	if err != nil {
+		glog.Errorf("Path %q fs stats error: %v", mountPath, err)
+		return
+	}
+	// actually if PV is provisioned by provisioner, the two values must be equal, but the PV may be
+	// created manually, so the PV capacity must not be greater than FS capacity
+	storage := pv.Spec.Capacity[v1.ResourceStorage]
+	if util.RoundDownCapacityPretty(capacityByte) < storage.Value() {
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	// TODO: make sure that PV used bytes is not greater that PV capacity ?
+
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+
+}
+
+func (monitor *LocalPVMonitor) checkPVAndBlockSize(mountPath string, pv *v1.PersistentVolume) {
+	capacityByte, err := monitor.VolUtil.GetBlockCapacityByte(mountPath)
+	if err != nil {
+		glog.Errorf("Path %q block stats error: %v", mountPath, err)
+		return
+	}
+	// actually if PV is provisioned by provisioner, the two values must be equal, but the PV may be
+	// created manually, so the PV capacity must not be greater than block device capacity
+	storage := pv.Spec.Capacity[v1.ResourceStorage]
+	if util.RoundDownCapacityPretty(capacityByte) < storage.Value() {
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	// TODO: make sure that PV used bytes is not greater that PV capacity ?
+
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+}
+
+// CheckNodeAffinity looks at the PV node affinity, and checks if the node has the same corresponding labels
+// This ensures that we don't mount a volume that doesn't belong to this node
+func CheckNodeAffinity(pv *v1.PersistentVolume, nodeLabels map[string]string) (bool, error) {
+	affinity, err := helper.GetStorageNodeAffinityFromAnnotation(pv.Annotations)
```

Comment:
this only get affinity from annotation?  what if this is using a beta version where affinity is part of the volume spec?

Reply:
good catch, i will update this